### PR TITLE
improve inspect.pyi

### PR DESCRIPTION
**Pull request by JelleZijlstra**
- Use TypeGuard for various is* functions (refer to #5406)
- Use collections.abc and builtin containers

### CODE REVIEW

File: stdlib/inspect.pyi [hunk 1]
```diff
@@ -324,12 +321,12 @@ CORO_SUSPENDED: str
...
 def getcoroutinestate(coroutine: object) -> str: ...
-def getgeneratorlocals(generator: Generator[Any, Any, Any]) -> Dict[str, Any]: ...
+def getgeneratorlocals(generator: Generator[Any, Any, Any]) -> dict[str, Any]: ...
 
 # TODO can we be more specific than "object"?
-def getcoroutinelocals(coroutine: object) -> Dict[str, Any]: ...
+def getcoroutinelocals(coroutine: object) -> dict[str, Any]: ...
 
-# Create private type alias to avoid conflict with symbol of same
+# Create private type alias to avoid confdict with symbol of same
```
Comment (srittau) [inline stdlib/inspect.pyi:329]: ^^

**Review by srittau — APPROVED** [review]
LGTM, except for pyright's warning and the marked spot.

File: stdlib/inspect.pyi (hunk 1 above)
Comment (JelleZijlstra) [reply stdlib/inspect.pyi]: Oops, thanks

**Comment by JelleZijlstra** [comment]
Looks like mypy crashes on `TypeGuard[Type[Any]]`. I'll investigate a bit more and probably report a mypy issue and for now keep `isclass` returning bool.

**Comment by JelleZijlstra** [comment]
Primer still fails to leave a comment.

**Comment by srittau** [comment]
I reset the permissions at https://github.com/python/typeshed/settings/actions and restarted the CI run.

**Comment by srittau** [comment]
And since that had no effect, I set the stricter permissions again. I'm trying something else.

**Comment by srittau** [comment]
🤞

**Comment by JelleZijlstra** [comment]
Well, it worked! But now i need to figure out what all those changes are about.

For reference: https://github.com/sphinx-doc/sphinx/blob/4.x/sphinx/util/inspect.py#L252, https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/elements/doc_string.py#L105

Looks like mypy has some bug around a TypeGuard anded with another if check.

**Comment by JelleZijlstra** [comment]
Made this include #5551 to see what happens with mypy-primer.

**Comment by JelleZijlstra** [comment]
The coverage.py one is still a bug in mypy (https://github.com/sphinx-doc/sphinx/blob/4.x/sphinx/ext/coverage.py#L218). It somehow narrows the type down from Any to object.

**Review by srittau — APPROVED** [review]
The impact of the mypy bug seems to be rather small. The problem with `inspect.py` is unfortunate, but a standard problem with mypy not understanding `hasattr()`.

**Pull request merged**